File: python/hard/muh_and_cubewalls.py
n,w=map(int,input().split())

a=list(map(int,input().split()))
b=list(map(int,input().split()))

if w==1:
    print(n)
    exit()

A=[]
B=[]

for i in range(1,n):
    A.append(a[i]-a[i-1])

for i in range(1,w):
    B.append(b[i]-b[i-1])

pi=[0]*len(B)

j=0

for i in range(1,len(B)):

    while j>0 and B[i]!=B[j]:
        j=pi[j-1]

    if B[i]==B[j]:
        j+=1

    pi[i]=j

ans=0
j=0

for x in A:

    while j>0 and x!=B[j]:
        j=pi[j-1]

    if x==B[j]:
        j+=1

    if j==len(B):
        ans+=1
        j=pi[j-1]

print(ans)

File: js/hard/muh-and-cubewalls.js
const fs=require("fs");

const input=fs.readFileSync(0,"utf8").trim().split(/\s+/).map(Number);

let p=0;

let n=input[p++];
let w=input[p++];

let a=[];
let b=[];

for(let i=0;i<n;i++)
a.push(input[p++]);

for(let i=0;i<w;i++)
b.push(input[p++]);

if(w===1){
    console.log(n);
    process.exit();
}

let A=[];
let B=[];

for(let i=1;i<n;i++)
A.push(a[i]-a[i-1]);

for(let i=1;i<w;i++)
B.push(b[i]-b[i-1]);

let pi=Array(B.length).fill(0);

for(let i=1,j=0;i<B.length;i++){

    while(j>0 && B[i]!==B[j])
        j=pi[j-1];

    if(B[i]===B[j])
        j++;

    pi[i]=j;
}

let ans=0;

for(let i=0,j=0;i<A.length;i++){

    while(j>0 && A[i]!==B[j])
        j=pi[j-1];

    if(A[i]===B[j])
        j++;

    if(j===B.length){
        ans++;
        j=pi[j-1];
    }
}

console.log(ans);